File: web/server/canvas/session.ts
import crypto from "node:crypto";
import type { ServerResponse } from "node:http";

import { logger } from "../utils/logger";
import { buildCanvasToolRequest } from "./operations";
import type { ToolName } from "./schemas";
import { compactCanvasState, compactNode, isToolName, parseToolInput } from "./tools";
import type { CanvasSnapshot } from "./types";

type PendingRequest = { clientId: string; resolve: (value: unknown) => void; reject: (error: Error) => void };

export const AGENT_PROTOCOL_VERSION = 6;

const SITE_TOOLS = new Set<ToolName>([
    "site_navigate",
    "canvas_list_projects",
    "assets_list",
    "assets_add",
    "generation_get_status",
]);
const READ_TOOLS = new Set<ToolName>(["canvas_get_state", "canvas_get_selection", "canvas_export_snapshot"]);

/** 管理网页画布连接、状态与工具请求。 */
export class CanvasSession {
    private clients = new Map<string, ServerResponse>();
    private clientFocusOrder = new Map<string, number>();
    private pending = new Map<string, PendingRequest>();
    private canvasStates = new Map<string, CanvasSnapshot>();
    private activeClientId = "";
    private boundClientId = "";
    private focusSequence = 0;

    /** 获取当前目标网页的画布状态。 */
    private get canvasState() {
        return this.clients.has(this.targetClientId) ? this.canvasStates.get(this.targetClientId) || null : null;
    }

    /** 获取当前绑定或最近激活的网页客户端。 */
    private get targetClientId() {
        return this.boundClientId || this.activeClientId;
    }

    /** 返回 Canvas MCP 当前连接状态。 */
    health() {
        return { ok: true, protocolVersion: AGENT_PROTOCOL_VERSION, hasCanvas: Boolean(this.canvasState), clients: this.clients.size };
    }

    /** 读取指定网页上报的画布，避免受最近焦点或其他标签页影响。 */
    canvasStateForClient(clientId: string) {
        return this.clients.has(clientId) ? this.canvasStates.get(clientId) || null : null;
    }

    /** 建立网页与 Canvas MCP 之间的 SSE 连接。 */
    openEvents(url: URL, res: ServerResponse) {
        const clientId = url.searchParams.get("clientId") || crypto.randomUUID();
        logger.info("SSE client connected", { clientId });
        res.writeHead(200, { "Content-Type": "text/event-stream", "Cache-Control": "no-cache", Connection: "keep-alive" });
        this.clients.set(clientId, res);
        if (!this.clientFocusOrder.has(clientId)) this.clientFocusOrder.set(clientId, 0);
        if (!this.activeClientId) {
            this.activeClientId = clientId;
            this.clientFocusOrder.set(clientId, ++this.focusSequence);
        }
        sendEvent(res, "hello", { ok: true, protocolVersion: AGENT_PROTOCOL_VERSION, clientId });
        const timer = setInterval(() => sendEvent(res, "ping", { time: Date.now() }), 15000);
        res.on("close", () => {
            clearInterval(timer);
            logger.info("SSE client disconnected", { clientId });
            if (this.clients.get(clientId) !== res) return;
            this.clients.delete(clientId);
            this.clientFocusOrder.delete(clientId);
            this.canvasStates.delete(clientId);
            this.pending.forEach((item, requestId) => {
                if (item.clientId !== clientId) return;
                this.pending.delete(requestId);
                item.reject(new Error("请求页面已断开"));
            });
            if (this.activeClientId === clientId) this.activeClientId = [...this.clients.keys()].sort((a, b) => (this.clientFocusOrder.get(b) || 0) - (this.clientFocusOrder.get(a) || 0))[0] || "";
        });
    }

    /** 保存指定网页上报的最新画布快照。 */
    updateState(body: unknown, clientId?: string) {
        const targetClientId = clientId || this.activeClientId;
        if (!targetClientId || !this.clients.has(targetClientId)) return;
        const state = { ...((body && typeof body === "object" && !Array.isArray(body) ? body : {}) as Record<string, unknown>), clientId: targetClientId } as CanvasSnapshot;
        this.canvasStates.set(targetClientId, state);
        logger.debug("Canvas state updated", { clientId: targetClientId, nodes: state.nodes?.length || 0, connections: state.connections?.length || 0 });
    }

    /** 将指定网页设为最近激活的工具目标。 */
    activateClient(clientId: string) {
        if (!this.clients.has(clientId)) throw new Error("当前网页未连接");
        this.activeClientId = clientId;
        this.clientFocusOrder.set(clientId, ++this.focusSequence);
        logger.debug("Canvas client activated", { clientId });
    }

    /** 将当前工具调用固定绑定到指定网页。 */
    bindClient(clientId: string) {
        if (!this.clients.has(clientId)) throw new Error("当前网页未连接");
        this.boundClientId = clientId;
        logger.debug("Canvas client bound", { clientId });
    }

    /** 解除当前工具调用的网页绑定。 */
    releaseClient(clientId: string) {
        if (this.boundClientId === clientId) this.boundClientId = "";
        logger.debug("Canvas client released", { clientId });
    }

    /** 接收网页返回的工具调用结果。 */
    resolveResult(clientId: string, body: { requestId?: string; error?: string; result?: unknown }) {
        const item = body.requestId ? this.pending.get(body.requestId) : null;
        if (!item || !body.requestId || item.clientId !== clientId) return false;
        this.pending.delete(body.requestId);
        logger.debug("Canvas tool result received", { clientId, requestId: body.requestId, error: body.error });
        body.error ? item.reject(new Error(body.error)) : item.resolve(body.result);
        return true;
    }

    /** 校验工具参数并将调用分派到当前目标网页。 */
    async callTool(name: unknown, rawInput: unknown) {
        if (!isToolName(name)) throw new Error(`未知工具：${String(name)}`);
        logger.info("MCP tool called", { name, targetClientId: this.targetClientId });
        const input = parseToolInput(name, rawInput) as Record<string, unknown>;
        if (SITE_TOOLS.has(name)) {
            if (!this.clients.size) throw new Error("当前没有已连接网页");
            return await this.requestCanvasTool(name, input);
        }
        if (READ_TOOLS.has(name) && (!this.clients.size || !this.canvasState)) throw new Error("当前没有已连接画布");
        if (name === "canvas_get_state" || name === "canvas_export_snapshot") return compactCanvasState(this.canvasState);
        if (name === "canvas_get_selection") {
            const ids = new Set(this.canvasState?.selectedNodeIds || []);
            return { nodes: (this.canvasState?.nodes || []).filter((node) => ids.has(node.id)).map(compactNode) };
        }
        if (!this.clients.size) throw new Error("当前没有已连接画布");
        const request = buildCanvasToolRequest(name, input, this.canvasState);
        return await this.requestCanvasTool(request.name, request.input);
    }

    /** 向目标网页发送工具请求并等待调用结果。 */
    private async requestCanvasTool(name: ToolName, input: Record<string, unknown>) {
        const requestId = crypto.randomUUID();
        const clientId = this.targetClientId;
        const client = this.clients.get(clientId);
        if (!client) throw new Error("当前没有已连接画布");
        sendEvent(client, "tool_call", { requestId, name, input });
        logger.debug("Canvas tool request sent", { requestId, name, clientId });
        return await new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                this.pending.delete(requestId);
                logger.warn("Canvas tool request timed out", { requestId, name, clientId });
                reject(new Error("画布操作超时"));
            }, 30000);
            this.pending.set(requestId, { clientId, resolve: (value) => (clearTimeout(timer), resolve(value)), reject: (error) => (clearTimeout(timer), reject(error)) });
        });
    }
}

/** 向 SSE 连接写入一个事件。 */
function sendEvent(res: ServerResponse, type: string, payload: unknown) {
    res.write(`event: ${type}\ndata: ${JSON.stringify(payload)}\n\n`);
}

File: mcp/src/canvas_mcp/session.py
"""Canvas session: SSE client registry, state cache and tool dispatch.

Port of ``web/server/canvas/session.ts``.
"""

from __future__ import annotations

import asyncio
import json
import uuid
from dataclasses import dataclass
from typing import Any

from canvas_mcp.config import AGENT_PROTOCOL_VERSION
from canvas_mcp.logger import logger
from canvas_mcp.operations import build_canvas_tool_request
from canvas_mcp.tools import compact_canvas_state, compact_node, is_tool_name, parse_tool_input
from canvas_mcp.types import CanvasSnapshot

SITE_TOOLS = frozenset({"site_navigate", "canvas_list_projects", "generation_get_status"})
READ_TOOLS = frozenset({"canvas_get_state", "canvas_get_selection", "canvas_export_snapshot"})

REQUEST_TIMEOUT_SECONDS = 30.0


@dataclass
class _PendingRequest:
    client_id: str
    future: asyncio.Future[Any]


def format_sse(event_type: str, payload: Any) -> str:
    """Serialise one server-sent event."""
    return f"event: {event_type}\ndata: {json.dumps(payload, ensure_ascii=False)}\n\n"


class CanvasSession:
    """Owns connected browser clients, their canvas snapshots and pending tool calls."""

    def __init__(self) -> None:
        self._clients: dict[str, asyncio.Queue[str]] = {}
        self._client_focus_order: dict[str, int] = {}
        self._pending: dict[str, _PendingRequest] = {}
        self._canvas_states: dict[str, CanvasSnapshot] = {}
        self._active_client_id = ""
        self._bound_client_id = ""
        self._focus_sequence = 0

    @property
    def _target_client_id(self) -> str:
        return self._bound_client_id or self._active_client_id

    @property
    def _canvas_state(self) -> CanvasSnapshot | None:
        target = self._target_client_id
        return self._canvas_states.get(target) if target in self._clients else None

    def health(self) -> dict[str, Any]:
        """Connection status returned by ``GET /health``."""
        return {
            "ok": True,
            "protocolVersion": AGENT_PROTOCOL_VERSION,
            "hasCanvas": bool(self._canvas_state),
            "clients": len(self._clients),
        }

    def canvas_state_for_client(self, client_id: str) -> CanvasSnapshot | None:
        """Read one client's snapshot without focus/binding influence."""
        return self._canvas_states.get(client_id) if client_id in self._clients else None

    def open_events(self, client_id: str) -> asyncio.Queue[str]:
        """Register an SSE client and return the queue its events are pushed to."""
        logger.info("SSE client connected", {"clientId": client_id})
        queue: asyncio.Queue[str] = asyncio.Queue()
        self._clients[client_id] = queue
        self._client_focus_order.setdefault(client_id, 0)
        if not self._active_client_id:
            self._active_client_id = client_id
            self._focus_sequence += 1
            self._client_focus_order[client_id] = self._focus_sequence
        return queue

    def close_events(self, client_id: str) -> None:
        """Drop a disconnected SSE client and reject its pending tool calls."""
        logger.info("SSE client disconnected", {"clientId": client_id})
        if client_id not in self._clients:
            return
        self._clients.pop(client_id, None)
        self._client_focus_order.pop(client_id, None)
        self._canvas_states.pop(client_id, None)
        for request_id, item in list(self._pending.items()):
            if item.client_id != client_id:
                continue
            self._pending.pop(request_id, None)
            if not item.future.done():
                item.future.set_exception(RuntimeError("请求页面已断开"))
        if self._active_client_id == client_id:
            remaining = sorted(self._clients, key=lambda cid: self._client_focus_order.get(cid, 0), reverse=True)
            self._active_client_id = remaining[0] if remaining else ""

    def update_state(self, body: Any, client_id: str | None = None) -> None:
        """Store the latest snapshot reported by a client."""
        target = client_id or self._active_client_id
        if not target or target not in self._clients:
            return
        state: CanvasSnapshot = {**(body if isinstance(body, dict) else {}), "clientId": target}
        self._canvas_states[target] = state
        logger.debug(
            "Canvas state updated",
            {"clientId": target, "nodes": len(state.get("nodes") or []), "connections": len(state.get("connections") or [])},
        )

    def activate_client(self, client_id: str) -> None:
        """Make a client the most recently focused tool target."""
        if client_id not in self._clients:
            raise ValueError("当前网页未连接")
        self._active_client_id = client_id
        self._focus_sequence += 1
        self._client_focus_order[client_id] = self._focus_sequence
        logger.debug("Canvas client activated", {"clientId": client_id})

    def bind_client(self, client_id: str) -> None:
        """Pin tool calls to one client."""
        if client_id not in self._clients:
            raise ValueError("当前网页未连接")
        self._bound_client_id = client_id
        logger.debug("Canvas client bound", {"clientId": client_id})

    def release_client(self, client_id: str) -> None:
        """Release the pinned client when it matches."""
        if self._bound_client_id == client_id:
            self._bound_client_id = ""
        logger.debug("Canvas client released", {"clientId": client_id})

    def resolve_result(self, client_id: str, body: Any) -> bool:
        """Deliver a browser tool result to the waiting request."""
        data = body if isinstance(body, dict) else {}
        request_id = data.get("requestId")
        item = self._pending.get(request_id) if isinstance(request_id, str) and request_id else None
        if not item or item.client_id != client_id or item.future.done():
            return False
        self._pending.pop(request_id, None)
        logger.debug("Canvas tool result received", {"clientId": client_id, "requestId": request_id, "error": data.get("error")})
        if data.get("error"):
            item.future.set_exception(RuntimeError(str(data["error"])))
        else:
            item.future.set_result(data.get("result"))
        return True

    async def call_tool(self, name: Any, raw_input: Any) -> Any:
        """Validate arguments and dispatch a tool call to the current target client."""
        if not is_tool_name(name):
            raise ValueError(f"未知工具：{name}")
        logger.info("MCP tool called", {"name": name, "targetClientId": self._target_client_id})
        input_data = parse_tool_input(name, raw_input)
        if name in SITE_TOOLS:
            if not self._clients:
                raise ValueError("当前没有已连接网页")
            return await self._request_canvas_tool(name, input_data)
        if name in READ_TOOLS and (not self._clients or not self._canvas_state):
            raise ValueError("当前没有已连接画布")
        if name in ("canvas_get_state", "canvas_export_snapshot"):
            return compact_canvas_state(self._canvas_state)
        if name == "canvas_get_selection":
            selected = set((self._canvas_state or {}).get("selectedNodeIds") or [])
            nodes = (self._canvas_state or {}).get("nodes") or []
            return {"nodes": [compact_node(node) for node in nodes if node.get("id") in selected]}
        if not self._clients:
            raise ValueError("当前没有已连接画布")
        request = build_canvas_tool_request(name, input_data, self._canvas_state)
        return await self._request_canvas_tool(request["name"], request["input"])

    def _send_event(self, client_id: str, event_type: str, payload: Any) -> None:
        queue = self._clients.get(client_id)
        if queue is not None:
            queue.put_nowait(format_sse(event_type, payload))

    async def _request_canvas_tool(self, name: str, input_data: dict[str, Any]) -> Any:
        """Relay a tool request to the target browser and await its result."""
        request_id = str(uuid.uuid4())
        client_id = self._target_client_id
        if client_id not in self._clients:
            raise ValueError("当前没有已连接画布")
        self._send_event(client_id, "tool_call", {"requestId": request_id, "name": name, "input": input_data})
        logger.debug("Canvas tool request sent", {"requestId": request_id, "name": name, "clientId": client_id})
        future: asyncio.Future[Any] = asyncio.get_running_loop().create_future()
        self._pending[request_id] = _PendingRequest(client_id, future)
        try:
            return await asyncio.wait_for(future, timeout=REQUEST_TIMEOUT_SECONDS)
        except TimeoutError:
            self._pending.pop(request_id, None)
            logger.warn("Canvas tool request timed out", {"requestId": request_id, "name": name, "clientId": client_id})
            raise ValueError("画布操作超时") from None
